File: frontend/src/api/correspondences.ts
import api from './client';
import type { CorrespondenceSystem, CorrespondenceAssignment, ResolvedCorrespondence } from '../types';

// === Correspondence Systems ===

export async function getCorrespondenceSystems(): Promise<CorrespondenceSystem[]> {
  const res = await api.get('/api/correspondence-systems');
  return res.data;
}

export async function getCorrespondenceSystem(systemId: number): Promise<CorrespondenceSystem & { assignments: CorrespondenceAssignment[] }> {
  const res = await api.get(`/api/correspondence-systems/${systemId}`);
  return res.data;
}

export async function createCorrespondenceSystem(data: {
  name: string;
  description?: string;
  cartomancy_type?: string;
  naming_style?: string;
}): Promise<{ id: number }> {
  const res = await api.post('/api/correspondence-systems', data);
  return res.data;
}

export async function updateCorrespondenceSystem(systemId: number, data: { name?: string; description?: string }) {
  await api.put(`/api/correspondence-systems/${systemId}`, data);
}

export async function deleteCorrespondenceSystem(systemId: number) {
  await api.delete(`/api/correspondence-systems/${systemId}`);
}

export async function cloneCorrespondenceSystem(systemId: number, name: string): Promise<{ id: number }> {
  const res = await api.post(`/api/correspondence-systems/${systemId}/clone`, { name });
  return res.data;
}

// === System Assignments ===

export async function getSystemAssignments(
  systemId: number,
  archetypeId?: number,
): Promise<CorrespondenceAssignment[]> {
  const params = archetypeId ? { archetype_id: archetypeId } : {};
  const res = await api.get(`/api/correspondence-systems/${systemId}/assignments`, { params });
  return res.data;
}

export async function bulkSetAssignments(
  systemId: number,
  assignments: { archetype_id: number; field_name: string; field_value: string }[],
  sourceGroup?: string,
) {
  await api.put(`/api/correspondence-systems/${systemId}/assignments`, {
    assignments,
    source_group: sourceGroup ?? null,
  });
}

export async function setAssignment(
  systemId: number,
  archetypeId: number,
  fieldName: string,
  value: string,
  sourceGroup?: string,
) {
  await api.put(`/api/correspondence-systems/${systemId}/assignments/${archetypeId}/${fieldName}`, {
    value,
    source_group: sourceGroup ?? null,
  });
}

export async function setAssignmentValues(
  systemId: number,
  archetypeId: number,
  fieldName: string,
  values: string[],
  sourceGroup?: string,
) {
  await api.put(`/api/correspondence-systems/${systemId}/assignments/${archetypeId}/${fieldName}`, {
    values,
    source_group: sourceGroup ?? null,
  });
}

export async function expandElementalZodiac(
  systemId: number,
  archetypeIds: number[],
  sourceGroup: string,
) {
  const res = await api.post(
    `/api/correspondence-systems/${systemId}/expand-elemental-zodiac`,
    { archetype_ids: archetypeIds, source_group: sourceGroup },
  );
  return res.data;
}

export async function deleteAssignment(
  systemId: number,
  archetypeId: number,
  fieldName: string,
  options?: { sourceGroup?: string; all?: boolean },
) {
  const params = new URLSearchParams();
  if (options?.sourceGroup) params.set('source_group', options.sourceGroup);
  if (options?.all) params.set('all', 'true');
  const qs = params.toString() ? `?${params}` : '';
  await api.delete(`/api/correspondence-systems/${systemId}/assignments/${archetypeId}/${fieldName}${qs}`);
}

// === Card-Level Overrides ===

export async function getCardCorrespondences(cardId: number): Promise<ResolvedCorrespondence[]> {
  const res = await api.get(`/api/cards/${cardId}/correspondences`);
  return res.data;
}

export async function setCardOverrides(
  cardId: number,
  overrides: { field_name: string; field_value?: string | null; field_values?: string[] }[],
) {
  await api.put(`/api/cards/${cardId}/correspondences`, { overrides });
}

export async function deleteCardOverride(cardId: number, fieldName: string) {
  await api.delete(`/api/cards/${cardId}/correspondences/${fieldName}`);
}

// === Deck-Level Overrides ===

export interface DeckCorrespondenceOverride {
  id: number;
  deck_id: number;
  archetype_id: number;
  archetype_name: string;
  cartomancy_type: string;
  rank: string | null;
  suit: string | null;
  card_type: string | null;
  field_name: string;
  field_value: string;
}

export async function getDeckCorrespondenceOverrides(deckId: number): Promise<DeckCorrespondenceOverride[]> {
  const res = await api.get(`/api/decks/${deckId}/correspondence-overrides`);
  return res.data;
}

export async function setDeckCorrespondenceOverride(
  deckId: number,
  archetypeId: number,
  fieldName: string,
  values: string[],
) {
  await api.put(`/api/decks/${deckId}/correspondence-overrides`, {
    archetype_id: archetypeId,
    field_name: fieldName,
    values,
  });
}

export async function deleteDeckCorrespondenceOverride(
  deckId: number,
  archetypeId: number,
  fieldName: string,
) {
  await api.delete(`/api/decks/${deckId}/correspondence-overrides/${archetypeId}/${fieldName}`);
}

// === Cross-System Queries ===

export async function getCorrespondencesByArchetype(archetypeId: number): Promise<CorrespondenceAssignment[]> {
  const res = await api.get(`/api/correspondences/by-archetype/${archetypeId}`);
  return res.data;
}

export async function compareCorrespondenceSystems(systemIds: number[]): Promise<CorrespondenceAssignment[]> {
  const res = await api.get('/api/correspondences/compare', {
    params: { systems: systemIds.join(',') },
  });
  return res.data;
}

// === Archetypes (for bulk editing) ===

export interface Archetype {
  id: number;
  name: string;
  cartomancy_type: string;
  rank: string | null;
  suit: string | null;
  card_type: string | null;
}

export async function getArchetypes(cartomancyType?: string): Promise<Archetype[]> {
  const params = cartomancyType ? { cartomancy_type: cartomancyType } : {};
  const res = await api.get('/api/archetypes', { params });
  return res.data;
}

// === Field Options ===

export interface FieldOption {
  id: number;
  field_name: string;
  option_value: string;
  sort_order: number;
}

export async function getFieldOptions(fieldName?: string): Promise<FieldOption[]> {
  const params = fieldName ? { field: fieldName } : {};
  const res = await api.get('/api/correspondence-field-options', { params });
  return res.data;
}

export async function addFieldOption(fieldName: string, optionValue: string): Promise<{ id: number }> {
  const res = await api.post('/api/correspondence-field-options', {
    field_name: fieldName,
    option_value: optionValue,
  });
  return res.data;
}

export async function updateFieldOption(optionId: number, data: { option_value?: string; sort_order?: number }) {
  await api.put(`/api/correspondence-field-options/${optionId}`, data);
}

export async function deleteFieldOption(optionId: number) {
  await api.delete(`/api/correspondence-field-options/${optionId}`);
}

export async function reorderFieldOptions(fieldName: string, orderedIds: number[]) {
  await api.post('/api/correspondence-field-options/reorder', {
    field_name: fieldName,
    ordered_ids: orderedIds,
  });
}

// === Card Names (for Reference tab) ===

export interface CardNameEntry {
  field_name: string;
  field_value: string;
  archetype: string | null;
  card_name: string;
  deck_name: string;
}

export async function getCardNames(): Promise<CardNameEntry[]> {
  const res = await api.get('/api/card-names');
  return res.data;
}

File: backend/routes/correspondences.py
"""
Correspondence system endpoints -- CRUD for systems, assignments, and card overrides.
"""

from flask import Blueprint, jsonify, request, current_app
from backend.utils import row_to_dict, require_json

correspondences_bp = Blueprint('correspondences', __name__)


# === Correspondence Systems ===

@correspondences_bp.route('/api/correspondence-systems')
def list_systems():
    db = current_app.config['DB']
    systems = db.get_correspondence_systems()
    return jsonify([row_to_dict(s) for s in systems])


@correspondences_bp.route('/api/correspondence-systems', methods=['POST'])
@require_json
def create_system(data):
    db = current_app.config['DB']
    name = data.get('name', '').strip()
    if not name:
        return jsonify({'error': 'Name is required'}), 400
    description = data.get('description', '').strip() or None
    try:
        system_id = db.create_correspondence_system(name, description)
    except Exception:
        return jsonify({'error': 'A system with that name already exists'}), 409
    return jsonify({'id': system_id}), 201


@correspondences_bp.route('/api/correspondence-systems/<int:system_id>')
def get_system(system_id):
    db = current_app.config['DB']
    system = db.get_correspondence_system(system_id)
    if not system:
        return jsonify({'error': 'System not found'}), 404
    result = row_to_dict(system)
    result['assignments'] = [
        row_to_dict(a) for a in db.get_system_assignments(system_id)
    ]
    return jsonify(result)


@correspondences_bp.route('/api/correspondence-systems/<int:system_id>', methods=['PUT'])
@require_json
def update_system(system_id, data):
    db = current_app.config['DB']
    system = db.get_correspondence_system(system_id)
    if not system:
        return jsonify({'error': 'System not found'}), 404
    try:
        db.update_correspondence_system(
            system_id,
            name=data.get('name'),
            description=data.get('description'),
        )
    except Exception:
        return jsonify({'error': 'A system with that name already exists'}), 409
    return jsonify({'ok': True})


@correspondences_bp.route('/api/correspondence-systems/<int:system_id>', methods=['DELETE'])
def delete_system(system_id):
    db = current_app.config['DB']
    system = db.get_correspondence_system(system_id)
    if not system:
        return jsonify({'error': 'System not found'}), 404
    db.delete_correspondence_system(system_id)
    return jsonify({'ok': True})


@correspondences_bp.route('/api/correspondence-systems/<int:system_id>/clone', methods=['POST'])
@require_json
def clone_system(system_id, data):
    db = current_app.config['DB']
    new_name = data.get('name', '').strip()
    if not new_name:
        return jsonify({'error': 'Name is required'}), 400
    new_id = db.clone_correspondence_system(system_id, new_name)
    if new_id is None:
        return jsonify({'error': 'Source system not found'}), 404
    return jsonify({'id': new_id}), 201


# === System Assignments ===

@correspondences_bp.route('/api/correspondence-systems/<int:system_id>/assignments')
def get_assignments(system_id):
    db = current_app.config['DB']
    archetype_id = request.args.get('archetype_id', type=int)
    assignments = db.get_system_assignments(system_id, archetype_id=archetype_id)
    return jsonify([row_to_dict(a) for a in assignments])


@correspondences_bp.route('/api/correspondence-systems/<int:system_id>/assignments', methods=['PUT'])
@require_json
def bulk_set_assignments(system_id, data):
    db = current_app.config['DB']
    assignments = data.get('assignments', [])
    source_group = data.get('source_group')  # None for manual edits
    if not assignments:
        return jsonify({'error': 'No assignments provided'}), 400
    try:
        db.bulk_set_system_assignments(system_id, assignments, source_group=source_group)
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    return jsonify({'ok': True})


@correspondences_bp.route(
    '/api/correspondence-systems/<int:system_id>/expand-elemental-zodiac',
    methods=['POST']
)
@require_json
def expand_elemental_zodiac_route(system_id, data):
    """Assign all 3 zodiac signs of each archetype's suit element.

    Body: { archetype_ids: [int], source_group: str }
    """
    db = current_app.config['DB']
    archetype_ids = data.get('archetype_ids', [])
    source_group = data.get('source_group') or ''
    if not archetype_ids or not source_group:
        return jsonify({'error': 'archetype_ids and source_group are required'}), 400
    report = db.expand_elemental_zodiac(system_id, archetype_ids, source_group)
    return jsonify({'ok': True, 'report': report})


@correspondences_bp.route(
    '/api/correspondence-systems/<int:system_id>/assignments/<int:archetype_id>/<field_name>',
    methods=['PUT']
)
@require_json
def set_assignment(system_id, archetype_id, field_name, data):
    db = current_app.config['DB']
    field_value = data.get('value', '').strip()
    source_group = data.get('source_group')
    if not field_value:
        return jsonify({'error': 'Value is required'}), 400
    try:
        db.set_system_assignment(
            system_id, archetype_id, field_name, field_value,
            source_group=source_group,
        )
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    return jsonify({'ok': True})


@correspondences_bp.route(
    '/api/correspondence-systems/<int:system_id>/assignments/<int:archetype_id>/<field_name>',
    methods=['DELETE']
)
def delete_assignment(system_id, archetype_id, field_name):
    """Delete assignment(s) for a cell.

    Query params:
        source_group: only delete the value from this specific group
        all: if 'true', delete all values (both manual and all groups)
    """
    db = current_app.config['DB']
    source_group = request.args.get('source_group')
    delete_all = request.args.get('all') == 'true'
    db.delete_system_assignment(
        system_id, archetype_id, field_name,
        source_group=source_group,
        delete_all_sources=delete_all,
    )
    return jsonify({'ok': True})


# === Card-Level Overrides ===

@correspondences_bp.route('/api/cards/<int:card_id>/correspondences')
def get_card_correspondences(card_id):
    db = current_app.config['DB']
    correspondences = db.get_card_correspondences(card_id)
    return jsonify(correspondences)


@correspondences_bp.route('/api/cards/<int:card_id>/correspondences', methods=['PUT'])
@require_json
def set_card_overrides(card_id, data):
    """Set one or more card-level overrides. Body: {overrides: [{field_name, field_value}]}"""
    db = current_app.config['DB']
    overrides = data.get('overrides', [])
    for override in overrides:
        field_name = override.get('field_name')
        field_value = override.get('field_value')
        if field_value is None or field_value == '':
            db.delete_card_correspondence_override(card_id, field_name)
        else:
            try:
                db.set_card_correspondence_override(card_id, field_name, field_value)
            except ValueError as e:
                return jsonify({'error': str(e)}), 400
    return jsonify({'ok': True})


@correspondences_bp.route('/api/cards/<int:card_id>/correspondences/<field_name>', methods=['DELETE'])
def delete_card_override(card_id, field_name):
    db = current_app.config['DB']
    db.delete_card_correspondence_override(card_id, field_name)
    return jsonify({'ok': True})


# === Cross-System Queries (for Reference tab) ===

@correspondences_bp.route('/api/correspondences/by-archetype/<int:archetype_id>')
def correspondences_by_archetype(archetype_id):
    db = current_app.config['DB']
    assignments = db.get_correspondences_by_archetype(archetype_id)
    return jsonify([row_to_dict(a) for a in assignments])


@correspondences_bp.route('/api/correspondences/compare')
def compare_systems():
    db = current_app.config['DB']
    system_ids_str = request.args.get('systems', '')
    try:
        system_ids = [int(x) for x in system_ids_str.split(',') if x.strip()]
    except ValueError:
        return jsonify({'error': 'Invalid system IDs'}), 400
    if not system_ids:
        return jsonify({'error': 'At least one system ID required'}), 400
    assignments = db.compare_correspondence_systems(system_ids)
    return jsonify([row_to_dict(a) for a in assignments])


@correspondences_bp.route('/api/correspondence-field-options')
def list_field_options():
    """Get all field options, optionally filtered by field name."""
    db = current_app.config['DB']
    field_name = request.args.get('field')
    options = db.get_field_options(field_name=field_name)
    return jsonify([row_to_dict(o) for o in options])


@correspondences_bp.route('/api/correspondence-field-options', methods=['POST'])
@require_json
def create_field_option(data):
    db = current_app.config['DB']
    field_name = data.get('field_name', '')
    option_value = (data.get('option_value', '') or '').strip()
    if not field_name or not option_value:
        return jsonify({'error': 'field_name and option_value are required'}), 400
    try:
        opt_id = db.add_field_option(field_name, option_value)
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    return jsonify({'id': opt_id}), 201


@correspondences_bp.route('/api/correspondence-field-options/<int:option_id>', methods=['PUT'])
@require_json
def update_field_option(option_id, data):
    db = current_app.config['DB']
    db.update_field_option(
        option_id,
        option_value=data.get('option_value'),
        sort_order=data.get('sort_order'),
    )
    return jsonify({'ok': True})


@correspondences_bp.route('/api/correspondence-field-options/<int:option_id>', methods=['DELETE'])
def delete_field_option(option_id):
    db = current_app.config['DB']
    db.delete_field_option(option_id)
    return jsonify({'ok': True})


@correspondences_bp.route('/api/correspondence-field-options/reorder', methods=['POST'])
@require_json
def reorder_field_options(data):
    db = current_app.config['DB']
    field_name = data.get('field_name')
    ordered_ids = data.get('ordered_ids', [])
    if not field_name:
        return jsonify({'error': 'field_name required'}), 400
    db.reorder_field_options(field_name, ordered_ids)
    return jsonify({'ok': True})


@correspondences_bp.route('/api/card-names')
def card_names():
    """Get card names in other languages, aggregated by archetype."""
    db = current_app.config['DB']
    cursor = db.conn.cursor()
    # Find all custom fields with "Name" in the field name
    cursor.execute('''
        SELECT ccf.field_name, ccf.field_value, c.archetype, c.name AS card_name,
               d.name AS deck_name
        FROM card_custom_fields ccf
        JOIN cards c ON c.id = ccf.card_id
        JOIN decks d ON d.id = c.deck_id
        WHERE ccf.field_name LIKE '%Name%'
          AND ccf.field_value IS NOT NULL
          AND ccf.field_value != ''
        ORDER BY c.archetype, ccf.field_name
    ''')
    rows = cursor.fetchall()
    return jsonify([row_to_dict(r) for r in rows])
